File: Data-Structure/Java/LeetCode/Binary Search/69. Sqrt(x).java
// /69. Sqrt(x)
/*
Given a non-negative integer x, compute and return the square root of x.

Since the return type is an integer, the decimal digits are truncated, and only the integer part of the result is returned.

Example 1: Input: x = 4 Output: 2
Example 2: Input: x = 8 Output: 2
Explanation: The square root of 8 is 2.82842..., and since the decimal part is truncated, 2 is returned.
*/
class Solution {
    public int mySqrt(int x) {
        int left = 0;
        int right = x;
        long num;
        int ans = -1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            num = (long) mid * mid; // need to convert to long
            if (num == x) {
                return mid;
            } else if (num < x) {
                ans = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return ans;
    }
}
/*
 * Search in Rotated Sorted Array II Perfect Squares Handshakes That Don't Cross
 */
//O(1)
class Solution {
    public int mySqrt(int x) {
        if (x < 2)
            return x;

        int left = (int) Math.pow(Math.E, 0.5 * Math.log(x));
        int right = left + 1;
        return (long) right * right > x ? left : right;
    }
}

File: Data-Structure/Nodejs/LeetCode by Category/Binary Search/69. Sqrt(x).js
// /69. Sqrt(x)
/*
Given a non-negative integer x, compute and return the square root of x.

Since the return type is an integer, the decimal digits are truncated, and only the integer part of the result is returned.

Example 1: Input: x = 4 Output: 2
Example 2: Input: x = 8 Output: 2
Explanation: The square root of 8 is 2.82842..., and since the decimal part is truncated, 2 is returned.
*/
{
    /**
 * @param {number} x
 * @return {number}
 */
    var mySqrt = function (x) {
        let left = 0;
        let right = x;
        let answer = -1;

        while (left <= right) {
            let mid = parseInt(left + ((right - left) >> 1));
            if (mid * mid === x) return mid;
            else if (mid * mid < x) {
                answer = mid; // take left side, the round
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return answer;
    };
    /*
    Search in Rotated Sorted Array II
    Perfect Squares
    Handshakes That Don't Cross
    */
}